src/commercetools/commercetools.service.ts

```diff
@@ -0,0 +1,409 @@
+    this.handlerOrderRedeem = handler;
+  }
+
+  private maxCartUpdateResponseTimeWithoutCheckingIfApiExtensionTimedOut: number =
+    getMaxCartUpdateResponseTimeWithoutCheckingIfApiExtensionTimedOut(
+      this.configService.get<number>(
+        'MAX_CART_UPDATE_RESPONSE_TIME_WITHOUT_CHECKING_IF_API_EXTENSION_TIMED_OUT',
+      ),
+    );
+
+  async handleCartUpdate(cart: CommerceToolsCart): Promise<{
+    actions: CartAction[];
+    status: boolean;
+  }> {
+    if (cart.version === 1) {
+      return await this.setCustomTypeForInitializedCart();
+    }
+
+    const actionBuilder = new ActionBuilder();
+    actionBuilder.setCart(cart);
+    actionBuilder.setCouponsLimit(
+      getCouponsLimit(
+        this.configService.get<number>('COMMERCE_TOOLS_COUPONS_LIMIT'),
+      ),
+    );
+
+    const cartDiscountApplyMode =
+      this.configService.get<string>(
+        'APPLY_CART_DISCOUNT_AS_CT_DIRECT_DISCOUNT',
+      ) === 'true'
+        ? CartDiscountApplyMode.DirectDiscount
+        : CartDiscountApplyMode.CustomLineItem;
+    actionBuilder.setCartDiscountApplyMode(cartDiscountApplyMode);
+
+    let taxCategory;
+    if (cartDiscountApplyMode === CartDiscountApplyMode.CustomLineItem) {
+      taxCategory = await this.getCouponTaxCategory(cart);
+    }
+    actionBuilder.setTaxCategory(taxCategory);
+
+    if (typeof this.handlerCartUpdate !== 'function') {
+      this.logger.error({
+        msg: `Error while commercetoolsService.validateCouponsAndPromotionsAndBuildCartActions handlerCartUpdate not configured`,
+      });
+      return {
+        status: false,
+        actions: [],
+      };
+    }
+    await this.handlerCartUpdate(
+      translateCtCartToCart(cart),
+      actionBuilder,
+      getPriceSelectorFromCtCart(cart),
+    );
+
+    const actions = actionBuilder.buildActions();
+
+    this.logger.debug({ msg: 'actions', actions });
+    return {
+      status: true,
+      actions: actions,
+    };
+  }
+
+  async handleAPIExtensionTimeout(
+    cart: CommerceToolsCart,
+    buildingResponseTime: number,
+  ) {
+    if (!cart?.custom?.fields?.discount_codes?.length) {
+      return;
+    }
+    if (
+      buildingResponseTime <
+      this.maxCartUpdateResponseTimeWithoutCheckingIfApiExtensionTimedOut
+    ) {
+      return;
+    }
+    let cartMutated = false;
+    for (let i = 0; i < 2; i++) {
+      await sleep(500);
+      const updatedCart = await this.commerceToolsConnectorService.findCart(
+        cart.id,
+      );
+      if (updatedCart.version > cart.version) {
+        cartMutated = true;
+        break;
+      }
+    }
+    if (cartMutated) {
+      return;
+    }
+    if (typeof this.handlerCartUpdate !== 'function') {
+      return this.logger.error({
+        msg: `Error while commercetoolsService.validateCouponsAndPromotionsAndBuildCartActions handlerCartUpdate not configured`,
+      });
+    }
+    await this.handlerCartUpdate(translateCtCartToCart(cart), null, null);
```

Comment:
changed to: `await this.CartUpdateHandler(translateCtCartToCart(cart)); //dropping sessions from coupons that are not included in cart (because of API extension timeout)`